redis/src/main/scala/akka/streams/alpakka/redis/model.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.streams.alpakka.redis
+
+import java.util.Optional
+import scala.collection.immutable.Seq
+import scala.util.{Failure, Success, Try}
+import scala.collection.JavaConverters._
+
+final class RedisFieldValue[K, V] private (val field: K, val value: V) {
+
+  override def equals(other: Any): Boolean = other match {
+    case that: RedisFieldValue[K, V] =>
+      field == that.field &&
+      value == that.value
+    case _ => false
+  }
+
+  override def hashCode(): Int =
+    31 * field.hashCode() + 31 * value.hashCode()
+
+  override def toString = s"RedisFieldValue($field, $value)"
+}
+
+object RedisFieldValue {
+
+  def create[K, V](key: K, value: V): RedisFieldValue[K, V] = new RedisFieldValue(key, value)
+  def apply[K, V](key: K, value: V): RedisFieldValue[K, V] = new RedisFieldValue(key, value)
+
+}
+
+final class RedisHMSet[K, V] private (val key: K, val values: Seq[RedisFieldValue[K, V]]) {
+
+  override def equals(other: Any): Boolean = other match {
+    case that: RedisHMSet[K, V] =>
+      key == that.key &&
+      values.length == that.values.length &&
+      values.zip(that.values).map { case (a, b) => a.equals(b) }.forall(_ == true)
+    case _ => false
+  }
+
+  override def hashCode(): Int =
+    values.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b) + 31 * key.hashCode()
+}
+
+object RedisHMSet {
+
+  def create[K, V](key: K, values: java.util.List[RedisFieldValue[K, V]]): RedisHMSet[K, V] =
+    new RedisHMSet[K, V](key, values.asScala.to[Seq])
+  def apply[K, V](key: K, values: Seq[RedisFieldValue[K, V]]): RedisHMSet[K, V] = new RedisHMSet[K, V](key, values)
+}
+
+final class RedisHSet[K, V] private (val key: K, val field: K, val value: V) {
+
+  override def equals(other: Any): Boolean = other match {
+    case that: RedisHSet[K, V] =>
+      key == that.key &&
+      field == that.field &&
+      value == that.value
+    case _ => false
+  }
+
+  override def hashCode(): Int =
+    31 * key.hashCode() + 31 * field.hashCode() + 31 * value.hashCode()
+}
+
+object RedisHSet {
+
+  def apply[K, V](key: K, field: K, value: V): RedisHSet[K, V] = new RedisHSet(key, field, value)
+  def create[K, V](key: K, field: K, value: V): RedisHSet[K, V] = new RedisHSet(key, field, value)
+}
+
+final class RedisKeyValue[K, V] private (val key: K, val value: V) {
+
+  override def equals(other: Any): Boolean = other match {
+    case that: RedisKeyValue[K, V] =>
+      key == that.key &&
+      value == that.value
+    case _ => false
+  }
+
+  override def hashCode(): Int =
+    31 * key.hashCode() + 31 * value.hashCode()
+}
```

Comment:
Unfortunately, supporting javadsl leads to some biolerplate code. Using case classes instead definitely would give as much more elegant code.

Reply:
You should still be able to use case classes while supporting javadsl. You just need to add the `create` method to the companion `object`. `apply`, `equals`, `hashCode` and `toString` can all be avoided :-) It is good to mark them `final`, too. You can see an example for such usages, here: https://github.com/akka/alpakka/blob/master/jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala#L65

But this is just a suggestion to keep the code smaller.

Reply:
The main motivation behind not using case classes is binary compatibility concerns. We can not add or remove attributes to case classes and keep binary compatible API. This is mostly a problem when case classes are used for settings, because settings have a higher probability of changing.

However for general classes like `RedisKeyValue` we could leave it as a case class reducing the needed boilerplate. Also the `create` method for Java API in the companion object is great and is needed even for a case class.